spec/controllers/links_controller_spec.rb

```diff
@@ -0,0 +1,241 @@
+      let(:user) { create :user }
+      before do
+        stub_current_user(user)
+        post :create, params: {
+          link: attributes_for(:link, vanity_string: 'test')
+        }
+      end
+
+      it 'redirects to home path' do
+        expect(response).to redirect_to home_path
+      end
+    end
+
+    context 'when user is not logged in' do
+      before do
+        post :create, params: {
+          link: attributes_for(:link, vanity_string: 'test')
+        }
+      end
+
+      it 'redirects to root path' do
+        expect(response).to redirect_to root_path
+      end
+    end
+
+    context 'when link has invalid parameters' do
+      before(:each) do
+        post :create, params: {
+          link: attributes_for(:link, full_url: nil)
+        }
+      end
+
+      it 'should not increase link count by 1' do
+        expect(assigns[:link].errors[:full_url]).to include "can't be blank"
+        expect(flash[:danger]).to be_present
+      end
+
+      it 'should return 302 status' do
+        expect(response.status).to eq(302)
+      end
+
+      it 'redirects to root path' do
+        expect(response).to redirect_to root_path
+      end
+    end
+  end
+
+  describe '#index' do
+    context 'when user is logged in' do
+      let(:user) { create :user }
+
+      before do
+        stub_current_user(user)
+        get :index
+      end
+
+      it 'assigns @links to current_user links' do
+        expect(assigns(:links)).to eq user.links
+      end
+
+      it 'renders the index template' do
+        expect(response).to render_template :index
+      end
+
+      it 'returns a 200 status' do
+        expect(response.status).to eq 200
+      end
+    end
+  end
+
+  describe '#show' do
+    let(:user) { create :user }
+
+    let(:link) { create :link }
+
+    before do
+      stub_current_user(user)
+      get :show, params: { id: link.id }
+    end
+
+    it 'renders show template' do
+      expect(response).to render_template :show
+    end
+
+    it 'returns a 200 status' do
+      expect(response.status).to eq 200
+    end
+  end
+
+  describe '#edit' do
+    let(:user) { create :user }
+
+    let(:link) { create :link }
+
+    before do
+      stub_current_user(user)
+      get :edit, params: { id: link.id }
+    end
+
+    it 'renders edit template' do
+      expect(response).to render_template :edit
+    end
+
+    it 'returns a 200 status' do
+      expect(response.status).to eq 200
+    end
+  end
+
+  describe '#update' do
+    let(:user) { create :user }
+
+    let!(:link) { create(:link, vanity_string: 'update') }
+
+    before do
+      stub_current_user(user)
+      put :update, params: { id: link.id, link: attributes_for(:link) }
+    end
+
+    it 'redirects to home path' do
+      expect(response).to redirect_to home_path
+    end
+
+    it 'returns a 302 status' do
+      expect(response.status).to eq 302
+    end
+
+    context 'when link parameters are valid' do
+      it 'updates the link' do
```

Comment:
You should actually verify that the link has the updated content/value instead of just the success message.